### ISSUE TITLE
Autoregressive or normalizing flow for density estimation

### ISSUE THREAD

**Issue by malihass** [pyro-ppl/numpyro#1881]
Hi there, 
I have question about the implementation of density estimation in numpyro.
* Context
I am working on a problem where I would like to calibrate uncertain parameters of a model. One thing I'd like to do is to construct a data-based prior distribution for these parameters, instead of using a more arbitrary prior.

* Tool I'm looking for
I'd like to train a data-based prior distribution and sample it during the my MCMC process. One (probably overkill) way is to use a normalizing flow for that. I have seen some examples for training normalizing flow in Pyro (https://pyro.ai/examples/normalizing_flows_i.html) and I am trying to do the same with Numpyro using AutoregressiveNN. I am struggling a bit to so as I could not find a way to 1) compute log probability 2) sample the AutoregressiveNN. 
Do you know how we could implement the same example as pyro in numpyro?

Attached is my unsuccessful attempt with AutoregressiveNN and I suspect that the logprob caclulation is what I am messing up.
```python
import numpy as np
import numpyro
import numpyro.distributions as dist
from numpyro.nn import AutoregressiveNN
from sklearn.datasets import make_circles
import jax
import jax.numpy as jnp
from jax import grad, jit
from jax import random
from jax.example_libraries import optimizers
import torch
import matplotlib.pyplot as plt

# Set random seed for reproducibility
numpyro.set_platform("cpu")
numpyro.set_host_device_count(1)
rng_key = random.PRNGKey(0)

# Generate double circle dataset
n_samples = 10000
X, y = make_circles(n_samples=n_samples, factor=0.5, noise=0.05)
X = jnp.array(X)  # Convert to JAX array
y = jnp.array(y)  # Convert labels to JAX array

# Initialize the optimizer
learning_rate = 0.0001
optimizer = optimizers.adam(learning_rate)
opt_init, opt_update, get_params = optimizers.adam(learning_rate)

# Create the autoregressive nn
base_dist = dist.Normal(jnp.zeros(2), jnp.ones(2))
nn = AutoregressiveNN(input_dim=2, hidden_dims=(32, 32))
init_fun, apply_fun = nn
_, initial_params = init_fun(rng_key, (2,))
opt_state = opt_init(initial_params)

def loss_fn(params, X):
    # Forward pass
    loc, scale = apply_fun(params, X)
    scale = jnp.exp(scale)

    # Log probability 
    log_prob_base = jnp.sum(base_dist.log_prob((X - loc) / scale), axis=-1)
    log_det_jacobian = jnp.sum(jnp.log(scale), axis=-1)
    log_prob_flow = log_prob_base + log_det_jacobian
    return -jnp.mean(log_prob_flow)


# Define the training loop
def train(X, y, opt, num_steps=1000):
    for step in range(num_steps):
        params = get_params(opt)
        loss, grads = jax.value_and_grad(loss_fn)(params, X)
        opt = opt_update(0, grads, opt)

        if step % 100 == 0:
            print(f"Step {step}, Loss: {loss}")

# Train
train(X, y, opt_state, num_steps=1000)

# sample
input_samp = jax.random.normal(rng_key, shape=(10000,2))
loc, scale = apply_fun(get_params(opt_state), input_samp)

# plot
fig = plt.figure()
plt.plot(X[:,0], X[:,1],'o', color='k', label="True")
plt.plot(loc[:,0], loc[:,1],'o', color='b', label="Generated")
plt.legend()
plt.show()
```

**Comment by fehiepsi** [comment]
Hi @malihass, we mainly use forum https://forum.pyro.ai/ for questions. Could you ask your question there?
